receiver/dockerstatsreceiver/docker.go

```diff
@@ -0,0 +1,253 @@
+	wg := sync.WaitGroup{}
+	for _, c := range containerList {
+		wg.Add(1)
+		go func(container dtypes.Container) {
+			if !dc.shouldBeExcluded(container.Image) {
+				if cnt, ok := dc.inspectedContainerIsOfInterest(ctx, container.ID); ok {
+					dc.persistContainer(cnt)
+				}
+			} else {
+				dc.logger.Debug(
+					"Not monitoring container per ExcludedImages",
+					zap.String("image", container.Image),
+					zap.String("id", container.ID),
+				)
+			}
+			wg.Done()
+		}(c)
+	}
+	wg.Wait()
+	return nil
+}
+
+// FetchContainerStatsAndConvertToMetrics will query the desired container stats and send
+// converted metrics to the results channel, since this is intended to be run in a goroutine.
+func (dc *DockerClient) FetchContainerStatsAndConvertToMetrics(
+	ctx context.Context,
+	container DockerContainer,
+	results chan<- Result,
+) {
+	dc.logger.Debug("Fetching container stats.", zap.String("id", container.ID))
+	statsCtx, cancel := context.WithTimeout(ctx, dc.config.Timeout)
+	containerStats, err := dc.client.ContainerStats(statsCtx, container.ID, false)
+	defer cancel()
+	if err != nil {
+		if docker.IsErrNotFound(err) {
+			dc.logger.Debug(
+				"Daemon reported container doesn't exist. Will no longer monitor.",
+				zap.String("id", container.ID),
+			)
+			dc.removeContainer(container.ID)
+		} else {
+			dc.logger.Warn(
+				"Could not fetch docker containerStats for container",
+				zap.String("id", container.ID),
+				zap.Error(err),
+			)
+		}
+
+		results <- Result{nil, err}
+		return
+	}
+
+	statsJSON, err := dc.toStatsJSON(containerStats, &container, results)
+	if err != nil { // results have been sent in converter
+		return
+	}
+
+	md, err := ContainerStatsToMetrics(statsJSON, &container, dc.config)
+	if err != nil {
+		dc.logger.Error(
+			"Could not convert docker containerStats for container id",
+			zap.String("id", container.ID),
+			zap.Error(err),
+		)
+		results <- Result{nil, err}
+		return
+	}
+
+	results <- Result{md, nil}
+}
+
+func (dc *DockerClient) toStatsJSON(
+	containerStats dtypes.ContainerStats,
+	container *DockerContainer,
+	results chan<- Result,
+) (*dtypes.StatsJSON, error) {
+	var statsJSON dtypes.StatsJSON
+	err := json.NewDecoder(containerStats.Body).Decode(&statsJSON)
+	containerStats.Body.Close()
+	if err != nil {
+		// EOF means there aren't any containerStats, perhaps because the container has been removed.
+		if err == io.EOF {
+			// It isn't indicative of actual error.
+			results <- Result{nil, nil}
+			return nil, err
+		}
+		dc.logger.Error(
+			"Could not parse docker containerStats for container id",
+			zap.String("id", container.ID),
+			zap.Error(err),
+		)
+		results <- Result{nil, err}
+		return nil, err
+	}
+	return &statsJSON, nil
+}
+
+// Queries inspect api and returns *ContainerJSON and true when container should be queried for stats,
+// nil and false otherwise.
+func (dc *DockerClient) inspectedContainerIsOfInterest(ctx context.Context, cid string) (*dtypes.ContainerJSON, bool) {
```

Comment:
Ok, that makes sense. Is `shouldBeExcluded` needed though -- it looks like `excludedImageMatcher` can't be nil.

Reply:
it's a pointer to a `StringMatcher` so it can be right, but isn't in practice thus far?  I'm not sure of what is idiomatic but this seems like a straightforward nil check.